db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java

```diff
@@ -2152,13 +2173,23 @@ private void pullOptimizableFromJoinOrder() throws StandardException{
         assignedTableMap.xor(pullMe.getReferencedTableMap());
     }
 
+    /* This check may have other side effects but we couldn't find
+     * a better way at this point. Problems:
+     * 1) The same statement, when executed at a different time (being
+     *    the first statement or executed after some other statements)
+     *    , may have 0 or 1 returned from lcc.getActivationCount().
+     * 2) Not all kinds of statement have a ResultDescription.
+     */
     private boolean isPreparedStatement() {
         LanguageConnectionContext lcc =
             (LanguageConnectionContext) ContextService.
                 getContextOrNull(LanguageConnectionContext.CONTEXT_ID);
         if (lcc == null || lcc.getActivationCount() < 1)
             return false;
-        return lcc.getLastActivation().getPreparedStatement() != null;
+        Activation ac = lcc.getLastActivation();
+        ResultDescription rd = ac.getResultDescription();
+        return ac.getPreparedStatement() != null && rd != null &&
```

Comment:
Nit: you could save a couple method calls by checking ac.getPreparedStatement() == null first, and returning right away if that is true.

Reply:
I think I get the point of this comment and the other about assertion is not removed. There are two `ResultDescription`s that we can get:
```
ac.getResultDescription();
```
and
```
ac.getPreparedStatement().getResultDescription();
```
At a quick glance, they return different objects. One is a member in `BaseActivation` and the other is a member in `GenericPreparedStatement`. Not sure if they eventually point to the same thing, though.